x/subkeys/module.go

```diff
@@ -0,0 +1,159 @@
+package subkeys
+
+import (
+	"encoding/json"
+
+	"github.com/gorilla/mux"
+	"github.com/spf13/cobra"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	tx "github.com/cosmos/cosmos-sdk/x/subkeys/internal/types/tx"
+)
+
+// TODO:
+// * docs
+// * periodic fee
+
+var (
+	_ module.AppModule      = AppModule{}
+	_ module.AppModuleBasic = AppModuleBasic{}
+)
+
+// AppModuleBasic defines the basic application module used by the delegation module.
+type AppModuleBasic struct{}
+
+// Name returns the delegation module's name.
+func (AppModuleBasic) Name() string {
+	return ModuleName
+}
+
+// RegisterCodec registers the delegation module's types for the given codec.
+func (AppModuleBasic) RegisterCodec(cdc *codec.Codec) {
+	RegisterCodec(cdc)
+	tx.RegisterCodec(cdc)
+}
+
+// DefaultGenesis returns default genesis state as raw bytes for the delegation
+// module.
+func (AppModuleBasic) DefaultGenesis() json.RawMessage {
+	return []byte("[]")
+}
+
+// ValidateGenesis performs genesis state validation for the delegation module.
+func (a AppModuleBasic) ValidateGenesis(bz json.RawMessage) error {
+	_, err := a.getValidatedGenesis(bz)
+	return err
+}
+
+func (a AppModuleBasic) getValidatedGenesis(bz json.RawMessage) (GenesisState, error) {
+	cdc := codec.New()
```

Comment:
Why not use `ModuleCdc`?

Reply:
I honestly don't like that design. I think singletons are bad.

I would prefer to accept a application-level cdc in the constructor, but that cannot be done (as the app cdc is created from the list of all AppModuleBasic)

Reply:
However, if this is the approach requested by the core team, I will adjust my code.

Reply:
Ideally we do it this way, the singleton has a purpose within modules. Mainly because developers can modify and/or replace the singleton in order to register their custom types in which case we need to have them included when handling genesis state.

Reply:
Done